### PR TITLE
CDRIVER-594 fix all the SS/SDAM timers

### PR DESCRIPTION
**Pull request by hanumantmk**
There's a bunch in here.

@ajdavis, mind taking a look to see if the logic is correct?

### CODE REVIEW

File: src/mongoc/mongoc-topology.c [hunk 1]
```diff
@@ -399,7 +433,7 @@ mongoc_topology_server_by_id (mongoc_topology_t *topology, uint32_t id)
  */
 
 void
-_mongoc_topology_request_scan (mongoc_topology_t *topology)
+mongoc_topology_request_scan (mongoc_topology_t *topology)
```
Comment (hanumantmk) [inline src/mongoc/mongoc-topology.c]: in retrospect, we should be able to call this from topology_select, but it's just not right (it shouldn't start the scan and shouldn't possibly block).

File: src/mongoc/mongoc-topology.c [hunk 2]
```diff
@@ -321,8 +349,10 @@ mongoc_topology_select (mongoc_topology_t         *topology,
                                                            error);
 
       if (! selected_server) {
-         /* TODO request an immediate topology check here */
-         r = mongoc_cond_timedwait (&topology->cond_client, &topology->mutex, expire_at - now);
+         topology->scan_requested = true;
```
Comment (hanumantmk) [inline src/mongoc/mongoc-topology.c]: we shouldn't have to do this.   I just got hung up on the fact that we have the mutex here.  It's fine to either have a version of request_scan which assumes we have the lock, or to drop the lock and use request scan, then resume it

File: src/mongoc/mongoc-thread-private.h [hunk 3]
```diff
@@ -41,11 +41,14 @@ mongoc_cond_timedwait (pthread_cond_t  *cond,
 {
    struct timespec to;
    struct timeval tv;
+   int64_t msec;
```
Comment (hanumantmk) [inline src/mongoc/mongoc-thread-private.h]: I really messed this up the first time 'round -_-;

File: src/mongoc/mongoc-topology-description.c [hunk 4]
```diff
@@ -445,7 +445,13 @@ mongoc_topology_description_select (mongoc_topology_description_t *topology,
    }
 
    if (topology->type == MONGOC_TOPOLOGY_SINGLE) {
-      return topology->servers->items[0].item;
+      sd = topology->servers->items[0].item;
```
Comment (hanumantmk) [inline src/mongoc/mongoc-topology-description.c]: I think it's better for server selection to say we don't have a server if we've never managed to ismaster it at all

File: src/mongoc/mongoc-topology-private.h [hunk 5]
```diff
@@ -59,7 +61,8 @@ typedef struct _mongoc_topology_t
 } mongoc_topology_t;
 
 mongoc_topology_t *
-mongoc_topology_new (const mongoc_uri_t *uri);
+mongoc_topology_new (const mongoc_uri_t *uri,
+                     bool                single_threaded);
```
Comment (hanumantmk) [inline src/mongoc/mongoc-topology-private.h]: I think it makes more sense for the topology to be created in single/multi threaded mode and for it to own it's background thread

File: src/mongoc/mongoc-topology-scanner-private.h [hunk 6]
```diff
@@ -44,6 +44,7 @@ typedef struct mongoc_topology_scanner_node
    mongoc_async_cmd_t             *cmd;
    mongoc_stream_t                *stream;
    int64_t                         timestamp;
+   int64_t                         last_used;
```
Comment (hanumantmk) [inline src/mongoc/mongoc-topology-scanner-private.h]: This is messy, but I'm not quite sure where else to hang it.  Maybe on the streams themselves?  Bleh

File: src/mongoc/mongoc-topology.c [hunk 7]
```diff
@@ -48,21 +54,17 @@ _mongoc_topology_scanner_cb (uint32_t      id,
 
    topology = data;
 
-   /* if multithreaded, driver may write through invalidate_server, so lock */
```
Comment (hanumantmk) [inline src/mongoc/mongoc-topology.c]: This is probably wrong to remove actually

**Comment by ajdavis** [comment]
LGTM.

**Pull request merged**